docs: add `dataProvider` section on new tutorial (#3305)
refinedev/refine: documentation/docs/tutorial/2-understanding-dataprovider/2-create-dataprovider.md

---
id: create-dataprovider
title: 3. Create Data Provider From Scratch
tutorial:
    prev: tutorial/understanding-dataprovider/swizzle
    next: tutorial/understanding-resources/index
---

## Introduction

> API usage and standards are very different, so data providers may not be suitable for you. 

Reviewer: ```suggestion
The standards and usage of APIs can vary significantly, so it is possible that refine's built-in data providers may not be suitable for your needs.
```

In this case, you will need to write your own Data Provider.

Data providers work with adapter system infrastructure. So they can communicate with REST, GraphQL, RPC and SOAP based APIs. You can use `fetch`, `axios`, `apollo-client` or any library for this communication.

> Now let's prepare a data provider that communicates with a REST API. We preferred `axios` as a client.

Reviewer: ```suggestion
Let's now create a data provider that communicates with a REST API. We preferred `axios` as a client.
```

> Let's start by creating a file like the one below. We will add other methods step by step.

Reviewer: ```suggestion
We will begin by creating a file, as shown below, and we will add additional methods to it as we proceed.
```

```ts title="src/data-provider.ts"
import { DataProvider } from "@pankod/refine-core";

export const dataProvider = (apiUrl: string): DataProvider => ({
  // Methods
});
```

> ## Methods

Reviewer: Under the Methods heading explain that What are you going to implement in this section.

Now we'll add....

### getList

`getList` method is used to sort, filter and paginate to get a list of resources.
It takes a `resource`, `sort`, `pagination` and `filters` as parameters should return `data` and `total`.

Let's assume the API we want to implement is as follows: 

```bash
[GET] https://api.fake-rest.refine.dev/posts?title_like=Arch&_sort=id&_order=desc&_limit=10&_page=2

HTTP/2 200
Content-Type: application/json
x-total-count: 22
access-control-expose-headers: X-Total-Count

[
  {
    "id": 930,
    "title": "Rerum id laborum architecto et rerum earum.",
    "slug": "et-voluptas-corporis",
    "category": {
      "id": 4
    }
    "status": "draft",
  },
  {
    "id": 892,
    "title": "Architecto officiis sint voluptatem modi.",
    "slug": "iusto-est-corrupti",
    "category": {
      "id": 1
    },
    "status": "rejected",
  }
]             
```

> 1. `resource` params implements the resource name. In our case, it is `posts`.

Reviewer: So what does this method do? It's gets the data and return json you shown above? Give an explanation please

       */
      if ("field" in filter) {
        const { field, operator, value } = filter;
        const mappedOperator = mapOperator(operator);
        queryFilters[`${field}${mappedOperator}`] = value;
      }
    });

    return queryFilters;
  };
  // highlight-end

  getList: async ({ resource, pagination, sort, filters }) => {
    const url = `${apiUrl}/${resource}`;

    const { current = 1, pageSize = 10 } = pagination ?? {};

    const query: {
      _start?: number;
      _end?: number;
      _sort?: string;
      _order?: string;
    } = {
      _start: (current - 1) * pageSize,
      _end: current * pageSize,
    };

    // it is not implemented because it does not support this API.
    if (sort && sort.length > 0) {
      query._sort = sort[0].field;
      query._order = sort[0].order;
    }

    // highlight-next-line
    const queryFilters = generateFilters(filters);

    // highlight-next-line
    const { data, headers } = await axios.get(`${url}?${stringify(query)}&${stringify(queryFilters)}`);

    const total = +headers["x-total-count"];

    return {
      data,
      total,
    };
  },
  ```

### create

> The `create` method creates a new record with the `resource` and `variables` methods.

Reviewer: ```suggestion
The `create` method creates a new record with the `resource` and `variables` parameters.
```

```ts title="src/data-provider.ts"
create: async ({ resource, variables }) => {
  const url = `${apiUrl}/${resource}`;

  const { data } = await axios.post(url, variables);

  return {
    data,
  };
},
```

### update

> The `update` method updates the record with the `resource`, `id` and `variables` methods.

Reviewer: ```suggestion
The `update` method updates the record with the `resource`, `id` and `variables` parameters.
```

  const { data } = await axios.delete(url, {
    data: variables,
  });

  return {
    data,
  };
}
```

### getOne
The `getOne` method gets the record with the `resource` and `id` methods.

```ts title="src/data-provider.ts"
getOne: async ({ resource, id }) => {
  const url = `${apiUrl}/${resource}/${id}`;

  const { data } = await axios.get(url);

  return {
    data,
  };
}
```

### getMany
The `getMany` method gets the records with the `resource` and `ids` methods.

```ts title="src/data-provider.ts"
getMany: async ({ resource, ids }) => {
  const { data } = await axios.get(
    `${apiUrl}/${resource}?${stringify({ id: ids })}`
  );

  return {
    data,
  };
}
```

### getApiUrl

The `getApiUrl` method returns the `apiUrl` value.

```ts title="src/data-provider.ts"
getApiUrl: () => apiUrl,
```

## Errors

When an error is returned from the API, **refine** must be extended from [HttpError](../../../../packages/core/src/interfaces/HttpError.ts) to handle it. 
Axios interceptor can be used to transform the error from response before Axios returns the response to your code. 
Interceptors are methods which are triggered before the main method. 

> We create an `axiosInstance`, define an `interceptors` to handle an error and export it. Write this in a `utility` file.

Reviewer: ```suggestion
In a `utility` file, create an `axiosInstance` and define an `interceptor` to handle errors. Then export it.
```

---
id: create-dataprovider
title: 3. Create Data Provider From Scratch
tutorial:
    prev: tutorial/understanding-dataprovider/swizzle
    next: tutorial/understanding-resources/index
---

## Introduction

> The standards and usage of APIs can vary significantly, so it is possible that refine's built-in data providers may not be suitable for your needs.

Reviewer 2: ```suggestion
The standards and usage of APIs can vary significantly, so it is possible that **refine's** built-in data providers may not be suitable for your needs.
```

In this case, you will need to write your own Data Provider.

Data providers work with adapter system infrastructure. So they can communicate with REST, GraphQL, RPC and SOAP based APIs. You can use `fetch`, `axios`, `apollo-client` or any library for this communication.

We will begin by creating a file, as shown below, and we will add additional methods to it as we proceed.

> Using `axios` as our HTTP client will allow us to make efficient and reliable HTTP requests to our server. Interceptors provide several benefits, such as centralized error handling, the ability to modify request or response data, and the ability to show global loading indicators. To get started, let's go ahead and install `axios` in our project.

Reviewer: ```suggestion
Using `axios` as our HTTP client will allow us to make efficient and reliable HTTP requests to our server. Interceptors provide several benefits, such as centralized error handling, the ability to modify request or response data, and the ability to show global loading indicators.  
To get started, let's go ahead and install `axios` in our project.
```

```ts
import { useDeleteMany } from "@pankod/refine-core";

const { mutate } = useDeleteMany();

mutate({
    resource: "posts",
    ids: [2, 3],
});
```
> [Refer to the useDeleteMany documentation for more information. &#8594](../../api-reference/core/hooks/data/useDeleteMany.md)

### updateMany

This method allows us to update multiple items in a resource. Implementation of this method is optional. If you don't implement it, refine will use `update` method to handle multiple requests.

```ts title="src/data-provider.ts"
updateMany: async ({ resource, ids, variables }) => {
  const url = `${apiUrl}/${resource}/bulk`;
  const { data } = await httpClient.patch(url, { ids, variables });

  return {
    data
  };
}
```

**refine** will consume this `updateMany` method using the `useUpdateMany` data hook.

```ts
import { useUpdateMany } from "@pankod/refine-core";

const { mutate } = useUpdateMany();

mutate({
    resource: "posts",
    ids: [1, 2],
    values: { status: "draft" },
});
```
> [Refer to the useUpdateMany documentation for more information. &#8594](../../api-reference/core/hooks/data/useUpdateMany.md)

## metaData Usage

> Sometimes you may want to send a custom parameter to APIs, for example a custom header parameter. You can use `metaData` for this. The sent parameter can be accessed directly from the data provider.

Reviewer: ```suggestion
When using APIs, you may wish to include custom parameters, such as a custom header. To accomplish this, you can utilize the `metaData` field, which allows the sent parameter to be easily accessed by the data provider.```

},
```

**Parameter Types**

> | Name      | Type         | Default |

Reviewer 3: Can we use docgen for the properties table? @aliemir @alicanerdurmaz
